src/Symfony/Component/Routing/Generator/UrlGenerator.php

```diff
@@ -27,60 +27,83 @@
  */
 class UrlGenerator implements UrlGeneratorInterface, ConfigurableRequirementsInterface
 {
-    private const QUERY_FRAGMENT_DECODED = [
-        // RFC 3986 explicitly allows those in the query/fragment to reference other URIs unencoded
-        '%2F' => '/',
-        '%3F' => '?',
-        // reserved chars that have no special meaning for HTTP URIs in a query or fragment
-        // this excludes esp. "&", "=" and also "+" because PHP would treat it as a space (form-encoded)
-        '%40' => '@',
-        '%3A' => ':',
+    // Character classes as defined in RFC 3986 2.2 and 3
+    // @see https://tools.ietf.org/html/rfc3986#page-22 ff
+
+    /**
+     * these chars are only sub-delimiters that have no predefined meaning and can therefore be used literally
+     * so URI producing applications can use these chars to delimit subcomponents in a path segment without being
+     * encoded for better readability.
+     *
+     * @see https://tools.ietf.org/html/rfc3986#page-11 2
+     */
+    protected const SUB_DELIMITERS = [
         '%21' => '!',
-        '%3B' => ';',
-        '%2C' => ',',
+        '%24' => '$',
+        '%26' => '&',
+        '%27' => '\'', // TODO: we may want an option to include this as original implementation claims usage in HTML, although this is wrong and htmlspecialchars et. al should be used here
```

Comment:
we should remove this from the list to preserve the encoding of single quotes
even though encoding this char is not required by the RFC, the URLs generated by Symfony should preserve this property of being safe from the pov of HTML and JavaScript escaping.
I would not make this configurable.